### ISSUE TITLE
from sometimes succeeds when daemon is not running

### ISSUE THREAD

**Issue by moskvich412** [tiian/flom#9]
I am not sure this is an issue. Could be my misunderstanding.

When I execute

flom -a myIP -p myPort -d 0  -o 10 -- ls

It sometimes succeeds without daemon running. If I execute it on some other machine, then this obviously fails because connection can not be established. But when executed on machine with myIP, then it sometimes succeeds. Other times it gives the same connection error as from another machine.

**Comment by tiian** [comment]
Supposing you are executing something like this

`flom -a 192.168.123.20 -p 1234 -d 0 -o 10 -- ls`

the behaviour you see is the consequence of the "-d 0" parameter.
If you set the trace as below, you can see it:

`
tiian@ubuntu1404-64:~$ export FLOM_TRACE_MASK=0x8

tiian@ubuntu1404-64:~$ flom -a 192.168.123.20 -p 1234 -d 0 -o 10 -- ls

2019-10-25 21:02:31.845304 [1517/0x205f400] flom_client_connect

2019-10-25 21:02:31.845387 [1517/0x205f400] flom_client_connect_tcp

2019-10-25 21:02:31.845461 [1517/0x205f400] flom_client_connect_tcp: connection failed, a new daemon can not be started because daemon lifespan is 0

2019-10-25 21:02:31.845476 [1517/0x205f400] flom_client_connect_tcp/excp=3/ret_cod=-104/errno=111

2019-10-25 21:02:31.845484 [1517/0x205f400] flom_client_connect/excp=1/ret_cod=-104/errno=111

flom_client_connect: ret_cod=-104 (ERROR: 'connect' function returned an error condition)
`

The most important message is "flom_client_connect_tcp: connection failed, a new daemon can not be started because daemon lifespan is 0".
This is typically used when:

1. you decide to start a daemon, typically without a real command like "true"
2. you access the started daemon by other flom commands that you don't want to start their own daemon.

In other words, flom starts a daemon by itself if you don't ask something different, and that daemon will disappear after some seconds. This is typically good enough if you want to protect a resource time to time.
Or you can define your own locking architecture with a "dedicated" daemon and one or more clients.

Hoping this description will help, please tell me if I can consider closed the issue.
K.R.
Ch.F.

**Comment by moskvich412** [comment]
I think I did not describe it well.

I would expect flom to fail every time because daemon is not running. But sometimes it succeeds when I run on the same computer as the supplied IP address. When I run on a computer with a different IP, then it always fails as I expect.

**Comment by tiian** [comment]
If it succeeds, for some reason, there's an available flom daemon answering.
Why there's one is something that requires analysis.
In this case I suggest you to run your flom commands enabling the trace feature. In file https://github.com/tiian/flom/blob/master/src/flom_trace.h you can look all the modules you can trace.
I suggest to start with GENERIC + CLIENT + DAEMON + CONNS with

export FLOM_TRACE_MASK=0x39

if they are not enough, you may add even MSG (very verbose) and TLS or TCP depending on your configuration.
Comparing the traces, you should be able to collect some more information.
K.R.
Ch.F.

**Comment by moskvich412** [comment]
I see what happens. In testing, I executed 

flom -a myIP -p myPort -o 10 -- ls

without "-d 0". This immediately lists files. This must also start daemon for few seconds. because soon after this 

flom -a myIP -p myPort -d 0 -o 10 -- ls

(with "-d 0") succeeds. But after some time (10 seconds?) I get connect error. Is that normal/expected?

**Comment by tiian** [comment]
Yes, this is the expected behaviour.
By default a short term (about 5 seconds) daemon is started: it will stay up and running for at least 5000ms (see _DEFAULT_DAEMON_LIFESPAN=5000 in configure.ac, but it can be configured as you prefer with flom.conf file).
Every incoming connection resets the timer.
After 5000ms of inactivity the daemon stops.
This avoid frequent daemon restart that would create conflicts in port binding (especially with TCP/IP ports).
If the default behaviour doesn't fit your use case, you can take a look to the link in documentation I posted above.
K.R.
Ch.F.

**Comment by moskvich412** [comment]
Thank you!

I am in a multi-user environment. I am not sysadmin. I must rely on defaults which I can only see via flom --help ( man page not installed) and web. I did not see "default daemon lifespan".

This is not how I intend to use it. Daemon will run. But I had to test various fail scenarios.

Will any user be able to shut down daemon with "-x"? Is there a way to limit this? (This is not the same topic, though.)

**Comment by tiian** [comment]
For config, you don't have to be admin, every user can create its own config file as explained here https://sourceforge.net/p/flom/wiki/Configuration/

Security is described here: https://sourceforge.net/p/flom/wiki/Security/

Using TLS and X.509 certificates you can guarantee only users with a valid certificate can access a specific flom daemon.
There's no a concept of "authorization", all the users have the same privileges: if one user can connect, he can stop the daemon.
